File: go/beautiful-arrangement.go
func validPosition(i int, perm_i int, perm []bool) bool {
    if perm[i] {
        return false
    }
    return (i % perm_i) == 0 || (perm_i % i) == 0
}

func countBeautifulArrangement(perm_i int, perm []bool) int {
    if perm_i == len(perm) {
        return 1
    }
    count := 0
    for i := 1; i < len(perm); i++ {
        if validPosition(i, perm_i, perm) {
            perm[i] = true
            count += countBeautifulArrangement(perm_i + 1, perm)
            perm[i] = false
        }
    }
    return count
}

func countArrangement(n int) int {
    perm := make([]bool, n + 1)
    return countBeautifulArrangement(1, perm)
}

File: python/beautiful-arrangement.py
class Solution:
    def is_valid_placement(perm_i, i, perm):
        if perm[i]:
            return False
        return perm_i % i == 0 or i % perm_i == 0

    def build_beautiful_permutation(perm_i, perm):
        if len(perm) == perm_i:
            return 1
        count = 0
        for i in range(1, len(perm)):
            if Solution.is_valid_placement(perm_i, i, perm):
                perm[i] = True
                count += Solution.build_beautiful_permutation(perm_i + 1, perm)
                perm[i] = False
        return count

    def countArrangement(self, n: int) -> int:
        perm = [False] * (n + 1)
        return Solution.build_beautiful_permutation(1, perm)
